### DISCUSSION TITLE
Visualizer - Controlling what you see?

### DISCUSSION THREAD

**Discussion by ryanmadam** [winder/Universal-G-Code-Sender#2242] [Help]
I'm coming from Fusion360 on Windows, so grabbing the camera control and scrolling around or zooming with the center mouse button/wheel is what I'm used to.

In the Visualizer, I can only zoom.  I can't scroll or drag the view to a different position.  I can tilt the view this way and that, but inevitably, I end up in a spot where I can't see what I want - zooming takes it off the screen and I see no way to scroll the display area up or down so I can still see the part I'm trying to zoom in on.

Am I doing something wrong or does it really not have any ability to scroll the display?

Secondarily, and maybe this should be a separate discussion, but how can I be sure which point the white cone is on when I select a spot and say "Jog to this location?"  

I'd like to be able to verify the position of my workpiece by jogging to certain features.  This is related to my first question, because without being able to drag or scroll the window, I can never get close enough to be sure I'm on the right corner or hole that I want to use to index my bit.

**Comment by breiler** [comment]
I started writing a couple of lines about how to orient the vizualiser here:
https://github.com/winder/Universal-G-Code-Sender/wiki/Usage#visualizer

The white cone will not snap to the loaded model, so it is basically just for eyeballing where to send the machine. This would be a nice improvement though. To be able to find a specific point in the model today - you would need to find the coordinate in the gcode and send it manually.

**Comment by ryanmadam** [comment]
Holy cow, I can't believe I never tried shift+left click.  That works great, but since 99% of everything in UGS can be done with just the mouse, I usually have my keyboard kinda far from my machine.  I scoot over to it when I need to enter a line # for "Run from..." or change jog setting.  I'll take a look at program preferences and try to figure it out for myself, but could that command be remapped to center mouse click?

Thanks for the reply, and thanks for all the work you've put into UGS.  It's still full of surprises - I just found out about the "Outline" command yesterday while searching for a way to pan the visualizer.  Previously I was raising the Z0 and running jobs to figure out the outline.  This is so much easier.

Ryan

**Reply by breiler** [reply]
There is no way to remap the controls, added a feature request here #2243 for us to work on further.

**Reply by ryanmadam** [reply]
I haven't done much with Java - but I do C# every day for work.  I could maybe take a look if you point me in the right direction - such as which file has the current mapping and where might be a good place to add an option to toggle between the current setting and using the center mouse button.

**Reply by breiler** [reply]
Here is the code that handles the visualizer inputs: 
https://github.com/winder/Universal-G-Code-Sender/blob/master/ugs-platform/ugs-platform-visualizer/src/main/java/com/willwinder/ugs/nbm/visualizer/RendererInputHandler.java#L159

Here are the options panel:
https://github.com/winder/Universal-G-Code-Sender/blob/master/ugs-platform/ugs-platform-visualizer/src/main/java/com/willwinder/ugs/nbm/visualizer/options/VisualizerOptions.java

I am not sure on how to do the button mapping here, but you might be able figure something out. =)

**Reply by ryanmadam** [reply]
Awesome.  Thanks.  I'll take a look at it.

Ryan

**Reply by Dwel** [reply]
I didnt even know there was a 'line # for "Run from...' Ive been starting jobs all over, which seems a waste of time since it carves nothing till it gets to the point it was. Thanks for mentioning that.
